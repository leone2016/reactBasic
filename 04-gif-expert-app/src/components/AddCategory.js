import React, { useState } from 'react'
import PropTypes from 'prop-types';

 const AddCategory = (props) => {
    const { setcategories } = props;
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);
        console.log('handleInputChange Called');
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        if( inputValue.trim().length > 2 ){
            setcategories(cat=> [ inputValue, ...cat]);
            setInputValue('');
        }
       
    }
    return (
        <form onSubmit={handleSubmit}>
            <p> {inputValue} </p>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}
AddCategory.prototype = {
    setcategories : PropTypes.func.isRequired
}


export default AddCategory;
