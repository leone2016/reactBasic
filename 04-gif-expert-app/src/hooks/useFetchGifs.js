import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    /**
     * useEffect: se utiliza para evitar que se realice una peticion fetch
     * cada que el estado del componente cambia, ademas
     * Los efectos NO puedes ser async 
     * SOLUCION, add in the [ category ]
     * 
     * React Hook useEffect has a missing dependency: 'getGifs'. 
     * Either include it or remove the dependency array  react-hooks/exhaustive-deps
     */
    useEffect(() => {
        getGifs(category).then(imgs => {
            setState({
                data: imgs,
                loading: false
            })
        });
    }, [category]);



    return state;
}
