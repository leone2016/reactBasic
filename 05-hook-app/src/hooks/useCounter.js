import { useState } from "react"
/**
 * TEst
 * @param {*} initialState 
 * @returns {state,
        increment,
        decrement,
        resetear}
 */
export const useCounter = (initialState = 10 ) => {
    const [state, setstate] = useState(initialState);

    const increment = (factor = 1)=>{
        setstate( state + factor);
    }
    const decrement = (factor = 1)=>{
        setstate( state - factor);
    }
    const resetear = ()=>{
        setstate( initialState );
    }

    //useCounter
    return {
        state,
        increment,
        decrement,
        resetear
    }
}
