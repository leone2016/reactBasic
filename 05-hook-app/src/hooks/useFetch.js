import  { useState, useEffect, useRef } from 'react'

export const useFetch = (url) => {

    const isMounted = useRef(true);
    const initialState = { data: null, loading:true, error: null};
    const [state, setstate] = useState(initialState);

    useEffect(() => {
        return () => {
            isMounted.current = false; // este cambio no realiza la renderizacion del componente, solo la referencia del mismo 
        }
    }, [])


    useEffect(() => {
        setstate(initialState);
        fetch(url)
        .then( resp => resp.json() )
        .then( data => {
            setTimeout(() =>{
                if(isMounted.current){
                    setstate({ 
                        loading: false,
                        error: null,
                        data
                    });
                }else{
                    console.warn('setState no se llamo');
                }
                

            },4000);
           
        })
    },[url]);

    return  state;
}
