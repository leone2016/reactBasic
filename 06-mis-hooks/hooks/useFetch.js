import  { useState, useEffect, useRef } from 'react'

export const useFetch = (url) => {

    const isMounted = useRef(true);
    const initialState = { data: null, loading:true, error: null};
    const [state, setstate] = useState(initialState);

    useEffect(() => {
        return () => {
            isMounted.current = false; // este cambio no realiza la renderizacion del componente, solo la referencia del mismo 
        }
    }, [])


    useEffect(() => {
        setstate(initialState);
        fetch(url)
        .then( resp => resp.json() )
        .then( data => {
           /*  setTimeout(() =>{ */
                if(isMounted.current){
                    setstate({ 
                        loading: false,
                        error: null,
                        data
                    });
                }
                

           /*  },4000); */
           
        }).catch(()=>{
            setstate({
                data: null, 
                loading: false, 
                error: 'No se pudo cargar la info'
            });
        })
    },[url]);

    return  state;
}
