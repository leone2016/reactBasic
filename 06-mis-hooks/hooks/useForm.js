import  { useState } from 'react'
/**
 * Permite manejar cualquier formulario
 * 
 * @param {*} initialState  un ojeto de cualquier tipo
 */
export const useForm = (initialState = {}) => {

    const [values, setvalues] = useState(initialState);

    const handleInputChange = ({ target }) => {
        setvalues({
            ...values,
            [target.name]: target.value
        })
    }

    const reset = () => {
        setvalues(initialState);
    }

    return [values, handleInputChange, reset ];

}
