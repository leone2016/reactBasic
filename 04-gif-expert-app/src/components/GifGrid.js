import React from 'react';
import {GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';
/* import { getGifs } from '../helpers/getGifs'; */


export const GifGrid = ({ category }) => {

    const {data: images,  loading } = useFetchGifs(category);

   /*  const [images, setImages] = useState([]) */

  

    return (
        <>
            <h3>{category}</h3>
            <div className="card-grid">

            {loading && <h4>C A R G A N D O . . .</h4>}
                <ol>
                   {
                        images.map(img => (
                            <GifGridItem
                                key={img.id}
                                {...img}
                            />
                        ))
                    }
                </ol>
           
            </div>
        </>
    )
}
