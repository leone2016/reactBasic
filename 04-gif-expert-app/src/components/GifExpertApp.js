import React, { useState } from 'react'
import AddCategory from './AddCategory';
import { GifGrid } from './GifGrid';

const GifExpertApp = props => {
    const initialState = ['One Punch'];
    const [categories, setcategories] = useState(initialState);

    const handleAdd = (category) => {
        /* setcategories( ['HunterHunter', ...categories]); */
    }
    return (
        <div>
            <h2>GifExpertApp</h2>
            <AddCategory setcategories={setcategories} />
            <hr />
            <button onClick={handleAdd}>Agregar</button>
            <ol>
                {
                    categories.map((cat) => (
                      <GifGrid
                      key={cat}
                      category={cat}
                      />
                    )

                    )
                }
            </ol>
        </div>
    )
}

export default GifExpertApp
