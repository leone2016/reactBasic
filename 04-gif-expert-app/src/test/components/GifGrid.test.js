import React from 'react';
import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');
describe('UNIT TEST GifGrid.test.js ', () => {
   const category = 'Goku';

   test('debe de mostrar correctamente', () => {
      useFetchGifs.mockReturnValue({
         data: [],
         loading: true
      });
      const wrapper = shallow(<GifGrid category={category} />);
      expect(wrapper).toMatchSnapshot();
   });

   test('debe mostrar items cuando se cargan imagenes useFetchGifs', () => {
      const gifs = [{
         id: 'ABC',
         url: 'https://localhost/cualquierCosa/test.jpg',
         title: 'Cualquier Cosa'
      }]
      useFetchGifs.mockReturnValue({
         data: gifs,
         loading: true
      });
      const wrapper = shallow(<GifGrid category={category} />);
      expect(wrapper).toMatchSnapshot();

      expect(wrapper).toMatchSnapshot();
   });
});
