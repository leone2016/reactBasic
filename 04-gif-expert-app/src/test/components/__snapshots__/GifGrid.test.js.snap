// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UNIT TEST GifGrid.test.js  debe de mostrar correctamente 1`] = `
<Fragment>
  <h3>
    Goku
  </h3>
  <div
    className="card-grid"
  >
    <h4>
      C A R G A N D O . . .
    </h4>
    <ol />
  </div>
</Fragment>
`;

exports[`UNIT TEST GifGrid.test.js  debe mostrar items cuando se cargan imagenes useFetchGifs 1`] = `
<Fragment>
  <h3>
    Goku
  </h3>
  <div
    className="card-grid"
  >
    <h4>
      C A R G A N D O . . .
    </h4>
    <ol>
      <GifGridItem
        id="ABC"
        key="ABC"
        title="Cualquier Cosa"
        url="https://localhost/cualquierCosa/test.jpg"
      />
    </ol>
  </div>
</Fragment>
`;

exports[`UNIT TEST GifGrid.test.js  debe mostrar items cuando se cargan imagenes useFetchGifs 2`] = `
<Fragment>
  <h3>
    Goku
  </h3>
  <div
    className="card-grid"
  >
    <h4>
      C A R G A N D O . . .
    </h4>
    <ol>
      <GifGridItem
        id="ABC"
        key="ABC"
        title="Cualquier Cosa"
        url="https://localhost/cualquierCosa/test.jpg"
      />
    </ol>
  </div>
</Fragment>
`;
